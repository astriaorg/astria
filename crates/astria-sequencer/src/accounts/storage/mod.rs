pub(super) mod keys;
mod values;

pub(crate) use values::Value;
pub(super) use values::{
    Balance,
    Nonce,
};
