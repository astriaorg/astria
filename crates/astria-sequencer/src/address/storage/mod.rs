pub(super) mod keys;
mod values;

pub(super) use values::AddressPrefix;
pub(crate) use values::Value;
