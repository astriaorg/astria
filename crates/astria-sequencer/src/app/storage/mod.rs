mod values;

pub(crate) use values::{
    BlockHeight,
    BlockTimestamp,
    ChainId,
    RevisionNumber,
    StorageVersion,
    Value,
};
