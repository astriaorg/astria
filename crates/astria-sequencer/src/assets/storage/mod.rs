pub(super) mod keys;
mod values;

pub(super) use values::TracePrefixedDenom;
pub(crate) use values::Value;
