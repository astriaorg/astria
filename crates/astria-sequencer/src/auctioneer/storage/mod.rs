pub(super) mod keys;
mod values;

pub(super) use values::EnshrinedAuctioneerEntry;
pub(crate) use values::Value;
