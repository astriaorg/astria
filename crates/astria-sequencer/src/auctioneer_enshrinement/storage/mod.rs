pub(super) mod keys;
mod values;