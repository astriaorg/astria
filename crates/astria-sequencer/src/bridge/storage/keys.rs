use astria_core::primitive::v1::RollupId;
use base64::{
    display::Base64Display,
    engine::general_purpose::URL_SAFE,
};

use crate::{
    accounts::AddressBytes,
    storage::keys::AccountPrefixer,
};

pub(in crate::bridge) const INIT_BRIDGE_ACCOUNT_BASE_FEE: &str = "bridge/init_account_fee";
pub(in crate::bridge) const BRIDGE_LOCK_BYTE_COST_MULTIPLIER: &str =
    "bridge/lock_byte_cost_multiplier";
pub(in crate::bridge) const BRIDGE_SUDO_CHANGE_FEE: &str = "bridge/sudo_change_fee";

pub(in crate::bridge) const BRIDGE_ACCOUNT_PREFIX: &str = "bridge/account/";
const BRIDGE_ACCOUNT_SUDO_PREFIX: &str = "bridge/sudo/";
const BRIDGE_ACCOUNT_WITHDRAWER_PREFIX: &str = "bridge/withdrawer/";

pub(in crate::bridge) const DEPOSITS_EPHEMERAL: &str = "bridge/deposits";
const DEPOSIT_PREFIX: &str = "bridge/deposit/";

/// Example: `bridge/account/gGhH....zZ4=/rollup_id`.
///                         |base64 chars|
pub(in crate::bridge) fn rollup_id<T: AddressBytes>(address: &T) -> String {
    format!(
        "{}/rollup_id",
        AccountPrefixer::new(BRIDGE_ACCOUNT_PREFIX, address)
    )
}

/// Example: `bridge/account/gGhH....zZ4=/asset_id`.
///                         |base64 chars|
pub(in crate::bridge) fn asset_id<T: AddressBytes>(address: &T) -> String {
    format!(
        "{}/asset_id",
        AccountPrefixer::new(BRIDGE_ACCOUNT_PREFIX, address)
    )
}

/// Example: `bridge/sudo/gGhH....zZ4=`.
///                      |base64 chars|
pub(in crate::bridge) fn bridge_account_sudo_address<T: AddressBytes>(address: &T) -> String {
    AccountPrefixer::new(BRIDGE_ACCOUNT_SUDO_PREFIX, address).to_string()
}

/// Example: `bridge/withdrawer/gGhH....zZ4=`.
///                            |base64 chars|
pub(in crate::bridge) fn bridge_account_withdrawer_address<T: AddressBytes>(address: &T) -> String {
    AccountPrefixer::new(BRIDGE_ACCOUNT_WITHDRAWER_PREFIX, address).to_string()
}

/// Example: `bridge/account/gGhH....zZ4=/withdrawal_event/<event id>`.
///                         |base64 chars|                |UTF-8 chars|
pub(in crate::bridge) fn bridge_account_withdrawal_event<T: AddressBytes>(
    address: &T,
    withdrawal_event_id: &str,
) -> String {
    format!(
        "{}/withdrawal_event/{}",
        AccountPrefixer::new(BRIDGE_ACCOUNT_PREFIX, address),
        withdrawal_event_id
    )
}

pub(in crate::bridge) fn deposit(block_hash: &[u8; 32], rollup_id: &RollupId) -> String {
    format!(
        "{DEPOSIT_PREFIX}{}/{rollup_id}",
        Base64Display::new(block_hash, &URL_SAFE),
    )
}

pub(in crate::bridge) fn last_transaction_id_for_bridge_account<T: AddressBytes>(
    address: &T,
) -> String {
    format!(
        "{BRIDGE_ACCOUNT_PREFIX}{}/last_tx",
        Base64Display::new(address.address_bytes(), &URL_SAFE)
    )
}

#[cfg(test)]
mod tests {
    use astria_core::primitive::v1::Address;

    use super::*;

    const COMPONENT_PREFIX: &str = "bridge/";

    fn address() -> Address {
        "astria1rsxyjrcm255ds9euthjx6yc3vrjt9sxrm9cfgm"
            .parse()
            .unwrap()
    }

    #[test]
    fn keys_should_not_change() {
        insta::assert_snapshot!(INIT_BRIDGE_ACCOUNT_BASE_FEE);
        insta::assert_snapshot!(BRIDGE_LOCK_BYTE_COST_MULTIPLIER);
        insta::assert_snapshot!(BRIDGE_SUDO_CHANGE_FEE);
        insta::assert_snapshot!(DEPOSITS_EPHEMERAL);
        insta::assert_snapshot!(rollup_id(&address()));
        insta::assert_snapshot!(asset_id(&address()));
        insta::assert_snapshot!(bridge_account_sudo_address(&address()));
        insta::assert_snapshot!(bridge_account_withdrawer_address(&address()));
        insta::assert_snapshot!(bridge_account_withdrawal_event(&address(), "the-event"));
        insta::assert_snapshot!(deposit(&[1; 32], &RollupId::new([2; 32])));
        insta::assert_snapshot!(last_transaction_id_for_bridge_account(&address()));
    }

    #[test]
    fn keys_should_have_component_prefix() {
        assert!(INIT_BRIDGE_ACCOUNT_BASE_FEE.starts_with(COMPONENT_PREFIX));
        assert!(BRIDGE_LOCK_BYTE_COST_MULTIPLIER.starts_with(COMPONENT_PREFIX));
        assert!(BRIDGE_SUDO_CHANGE_FEE.starts_with(COMPONENT_PREFIX));
        assert!(DEPOSITS_EPHEMERAL.starts_with(COMPONENT_PREFIX));
        assert!(rollup_id(&address()).starts_with(COMPONENT_PREFIX));
        assert!(asset_id(&address()).starts_with(COMPONENT_PREFIX));
        assert!(bridge_account_sudo_address(&address()).starts_with(COMPONENT_PREFIX));
        assert!(bridge_account_withdrawer_address(&address()).starts_with(COMPONENT_PREFIX));
        assert!(
            bridge_account_withdrawal_event(&address(), "the-event").starts_with(COMPONENT_PREFIX)
        );
        assert!(deposit(&[1; 32], &RollupId::new([2; 32])).starts_with(COMPONENT_PREFIX));
        assert!(last_transaction_id_for_bridge_account(&address()).starts_with(COMPONENT_PREFIX));
    }

    #[test]
    fn bridge_account_prefix_should_be_prefix_of_relevant_keys() {
        assert!(rollup_id(&address()).starts_with(BRIDGE_ACCOUNT_PREFIX));
        assert!(asset_id(&address()).starts_with(BRIDGE_ACCOUNT_PREFIX));
        assert!(
            bridge_account_withdrawal_event(&address(), "the-event")
                .starts_with(BRIDGE_ACCOUNT_PREFIX)
        );
        assert!(
            last_transaction_id_for_bridge_account(&address()).starts_with(BRIDGE_ACCOUNT_PREFIX)
        );
    }
}
