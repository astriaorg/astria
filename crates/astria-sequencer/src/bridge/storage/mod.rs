mod values;

pub(crate) use values::{
    AddressBytes,
    BlockHeight,
    Deposits,
    Fee,
    IbcPrefixedDenom,
    RollupId,
    TransactionId,
    Value,
};
