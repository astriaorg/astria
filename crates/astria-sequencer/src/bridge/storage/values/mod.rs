mod address_bytes;
mod block_height;
mod deposits;
mod ibc_prefixed_denom;
mod rollup_id;
mod transaction_id;

use borsh::{
    BorshDeserialize,
    BorshSerialize,
};

pub(in crate::bridge) use self::{
    address_bytes::AddressBytes,
    block_height::BlockHeight,
    deposits::Deposits,
    ibc_prefixed_denom::IbcPrefixedDenom,
    rollup_id::RollupId,
    transaction_id::TransactionId,
};

#[derive(Debug, BorshSerialize, BorshDeserialize)]
pub(crate) struct Value<'a>(ValueImpl<'a>);

#[derive(Debug, BorshSerialize, BorshDeserialize)]
enum ValueImpl<'a> {
    RollupId(RollupId<'a>),
    IbcPrefixedDenom(IbcPrefixedDenom<'a>),
    AddressBytes(AddressBytes<'a>),
    BlockHeight(BlockHeight),
    Deposits(Deposits<'a>),
    TransactionId(TransactionId<'a>),
}

#[cfg(test)]
mod tests {
    use std::mem::discriminant;

    use astria_core::{
        crypto::ADDRESS_LENGTH,
        primitive::v1::{
            asset::IbcPrefixed as DomainIbcPrefixed,
            RollupId as DomainRollupId,
            TransactionId as DomainTransactionId,
            ROLLUP_ID_LEN,
            TRANSACTION_ID_LEN,
        },
        sequencerblock::v1::block::Deposit as DomainDeposit,
    };
    use insta::assert_snapshot;

    use super::*;

    #[test]
    fn value_impl_rollup_id_discriminant_unchanged() {
        assert_snapshot!(
            "value_impl_rollup_id_discriminant",
            format!(
                "{:?}",
                discriminant(&ValueImpl::RollupId(
                    (&DomainRollupId::new([0; ROLLUP_ID_LEN])).into()
                ))
            )
        );
    }

    #[test]
    fn value_impl_ibc_prefixed_denom_discriminant_unchanged() {
        assert_snapshot!(
            "value_impl_ibc_prefixed_denom_discriminant",
            format!(
                "{:?}",
                discriminant(&ValueImpl::IbcPrefixedDenom(
                    (&DomainIbcPrefixed::new([0; 32])).into()
                ))
            )
        );
    }

    #[test]
    fn value_impl_address_bytes_discriminant_unchanged() {
        assert_snapshot!(
            "value_impl_address_bytes_discriminant",
            format!(
                "{:?}",
                discriminant(&ValueImpl::AddressBytes((&[0; ADDRESS_LENGTH]).into()))
            )
        );
    }

    #[test]
    fn value_impl_block_height_discriminant_unchanged() {
        assert_snapshot!(
            "value_impl_block_height_discriminant",
            format!("{:?}", discriminant(&ValueImpl::BlockHeight(1.into())))
        );
    }

    #[test]
    fn value_impl_deposits_discriminant_unchanged() {
        assert_snapshot!(
            "value_impl_deposits_discriminant",
            format!(
                "{:?}",
                discriminant(&ValueImpl::Deposits(
                    Vec::<DomainDeposit>::new().iter().into()
                ))
            )
        );
    }

    #[test]
    fn value_impl_transaction_id_discriminant_unchanged() {
        assert_snapshot!(
            "value_impl_transaction_id_discriminant",
            format!(
                "{:?}",
                discriminant(&ValueImpl::TransactionId(
                    (&DomainTransactionId::new([0; TRANSACTION_ID_LEN])).into()
                ))
            )
        );
    }

    // Note: This test must be here instead of in `crate::storage` since `ValueImpl` is not
    // re-exported.
    #[test]
    fn stored_value_bridge_discriminant_unchanged() {
        use crate::storage::StoredValue;
        assert_snapshot!(
            "stored_value_bridge_discriminant",
            format!(
                "{:?}",
                discriminant(&StoredValue::Bridge(Value(ValueImpl::BlockHeight(
                    1.into()
                ))))
            )
        );
    }
}
