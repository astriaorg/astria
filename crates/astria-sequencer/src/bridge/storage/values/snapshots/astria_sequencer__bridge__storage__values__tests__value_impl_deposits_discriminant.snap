---
source: crates/astria-sequencer/src/bridge/storage/values/mod.rs
expression: "format!(\"{:?}\",\ndiscriminant(&ValueImpl::Deposits(Vec::<DomainDeposit>::new().iter().into())))"
---
Discriminant(4)
